### DISCUSSION TITLE
S3 fileupload with ACL public-read access

### DISCUSSION THREAD

**Discussion by UtkuDalmaz** [filamentphp/filament#11486] [Help]
### Package

Form builder

### Package Version

v3.2.34

### How can we help you?

When it uploads to S3, the file is being created as private. I want it to be a public file just like making it in the AWS panel with the option "make public using ACL." When using Laravel's filesystem (in a controller using putFileAs), it uploads public-read access properly, but in filament, it is uploaded as private, and I need to go and make it public using the "make public using ACL" button.

                    ->disk('s3')
                    ->directory('venues/')
                    ->imageResizeTargetWidth('700')
                    ->imageResizeMode('cover')
                    ->visibility('public')
                    ->image()
                    ->acceptedFileTypes(['image/jpeg'])
                    ->moveFiles()
                    ->getUploadedFileNameForStorageUsing(
                        fn (TemporaryUploadedFile $file): string => (string) self::$filename,
                    )
                    ->image()

here is the conf. 

Any idea how to make it work?

**Comment by brentkelly** [comment]
I'm not 100% sure this is the same issue, but I will leave this here in case others are running into an issue where S3 Uploads are simply not working using the default `FileUpload` component.

By default livewire attempts to use S3 ACL if you set to public `visibility`. This will throw an error unless your S3 bucket is configured to use ACLs - something that S3 recommends against as the default. The trouble is - by default laravel is configured to suppress errors for s3 - so your upload will silently fail and you will be left scratching your head mouthing a quiet WT... 😕🤷. (To change that, set `throw` to `true` in your `config/filesystems.php`.

If you set visiblity to `private`, livewire will use temporaryUrls for reading images which isn't really what we're after either. So ... checkmate.

The good news is there is a bit of a dirty hack workaound: setting this to neither public nor private allows you to treat uploads as public, but without trying to force file-specific ACL - instead relying on your bucket policy (to allow `s3:GetObject`). 

This is pretty much what I'm always after so in my Filament projects that use S3 I create my own `FileUpload` component that inherits from the base filament one:

```php
use Closure;
use Filament\Forms\Components\FileUpload as BaseFileUpload;

class FileUpload extends BaseFileUpload
{
    /**
     * By default livewire attempts to use S3 ACL if you set to public visibility. This
     * will throw an error unless your S3 bucket is configured to use ACLs - something
     * that S3 recommends against as the default. If you set visiblity to private,
     * livewire will use temporaryUrls for reading images. Setting this to neither
     * public or private allows you to treat uploads as public, but without trying
     * to force file-specific ACL - instead relying on your bucket policy (to allow
     * s3:GetObject)
     */
    protected string|Closure $visibility = 'bucket-defaults';
}
```

`bucket-defaults` is just any random string. It can even be an empty string - so long as it isn't `public` or `private`.

If I use that component in place of the Filament `FileUpload`, everything now works as I would expect.

**Reply by damilareonifade** [reply]
This does not work for me I have changed the visibility to multiple things I have cleared cache and I am still getting the same issue

**Reply by gmroughton** [reply]
Thank you for explaining why setting the visibility to something other than public / private works, I had found other solutions online that said to change it to 'publico' because they thought there was some kind of typo in the source code but I couldn't find anything to support that theory in the vendor files. This make so much more sense and now feels intentional vs a mistake in our codebase.

**Reply by brentkelly** [reply]
> This does not work for me I have changed the visibility to multiple things I have cleared cache and I am still getting the same issue

@damilareonifade Sorry just saw this - hope you have fixed the issue. I would suggest in that scenario to ensure you have changed your settings in `config/filesystems.php` to throw errors for s3 - which should then allow you to copy & paste the error into AI & get some suggestions to resolve.
